pkg/reconciler/broker/broker_test.go

```diff
@@ -691,7 +691,68 @@ func TestReconcile(t *testing.T) {
 			WantPatches: []clientgotesting.PatchActionImpl{
 				patchFinalizers(testNS, brokerName),
 			},
-		}, {
+		},
+		{
+			Name: "Successful Reconciliation, matching BrokerClass",
+			Key:  testKey,
+			Objects: []runtime.Object{
+				NewBroker(brokerName, testNS,
+					WithBrokerChannel(channel()),
+					WithInitBrokerConditions,
+					WithBrokerClass(eventing.ChannelBrokerClassValue)),
+				createChannel(testNS, triggerChannel, true),
+				NewDeployment(filterDeploymentName, testNS,
+					WithDeploymentOwnerReferences(ownerReferences()),
+					WithDeploymentLabels(resources.FilterLabels(brokerName)),
+					WithDeploymentServiceAccount(filterSA),
+					WithDeploymentContainer(filterContainerName, filterImage, livenessProbe(), readinessProbe(), envVars(filterContainerName), containerPorts(8080))),
+				NewService(filterServiceName, testNS,
+					WithServiceOwnerReferences(ownerReferences()),
+					WithServiceLabels(resources.FilterLabels(brokerName)),
+					WithServicePorts(servicePorts(8080))),
+				NewDeployment(ingressDeploymentName, testNS,
+					WithDeploymentOwnerReferences(ownerReferences()),
+					WithDeploymentLabels(resources.IngressLabels(brokerName)),
+					WithDeploymentServiceAccount(ingressSA),
+					WithDeploymentContainer(ingressContainerName, ingressImage, livenessProbe(), nil, envVars(ingressContainerName), containerPorts(8080))),
+				NewService(ingressServiceName, testNS,
+					WithServiceOwnerReferences(ownerReferences()),
+					WithServiceLabels(resources.IngressLabels(brokerName)),
+					WithServicePorts(servicePorts(8080))),
+			},
+			WantStatusUpdates: []clientgotesting.UpdateActionImpl{{
+				Object: NewBroker(brokerName, testNS,
+					WithBrokerChannel(channel()),
+					WithBrokerReady,
+					WithBrokerTriggerChannel(createTriggerChannelRef()),
+					WithBrokerAddress(fmt.Sprintf("%s.%s.svc.%s", ingressServiceName, testNS, utils.GetClusterDomainName())),
+					WithBrokerClass(eventing.ChannelBrokerClassValue),
+				),
+			}},
+			WantEvents: []string{
+				finalizerUpdatedEvent,
+			},
+			WantPatches: []clientgotesting.PatchActionImpl{
+				patchFinalizers(testNS, brokerName),
+			},
+		},
+		{
+			Name: "Successful Reconciliation, broker ignored because mismatching BrokerClass",
+			Key:  testKey,
+			Objects: []runtime.Object{
+				NewBroker(brokerName, testNS,
+					WithBrokerChannel(channel()),
+					WithInitBrokerConditions,
+					WithBrokerClass("broker-class-mismatch")),
+			},
+			WantEvents: []string{
+				finalizerUpdatedEvent,
+			},
+			WantPatches: []clientgotesting.PatchActionImpl{
```

Comment:
I think this is actually going to be a problem, since if it is not ours, we should not be adding a finalizer to it. We want it to implement Finalize, but only should be adding finalizers if it's ours. I don't see a way with genreconciler to avoid this behaviour however.

Reply:
This is a bit hacky but we could apply same filter in the `FinalizeKind`  method and skip the actual Finalizer logic. We can follow up on a cleaner way to do this.

Reply:
Actually this doesn't work since all controllers generated by genreconciler use the same Finalizer name. I think we should probably allow different Finalizer names.

Reply:
There are so many places where we enqueue the resource key to the workqueue. It's unlikely to apply filters to prevent keys being added to the queue.

Instead,  what do you think about adding a `Filter` interface in the genreconciler? When the actual reconciler implements this method, we call `Filter` and skip reconciliation and finalizer logic if the object doesn't pass.

Reply:
We should be able to use this.
https://github.com/knative/pkg/tree/master/injection#annotation-based-class-filters

Reply:
This is nice! But then we lose backwards compatibility. Our controller won't pick up brokers without the annotation. We need to somehow "backfill" brokers with default annotation.

Reply:
I thought there's a way to specify a default if it's missing? no?

Reply:
If you are referring to the genreconciler, I don't think there is an option:
https://github.com/knative/pkg/blob/a572b9922a9e2bdf860a9c9f3482782a912d9b25/codegen/cmd/injection-gen/generators/reconciler_reconciler.go#L266

Reply:
https://github.com/knative/eventing/pull/2749

Reply:
The link I sent earlier showed the mods you have to make to the broker_types.go so the right things happen. The magic is here:
https://github.com/knative/eventing/pull/2749/files#diff-1ad5ce06d4ca2d8245aa2beb89be5c87R33

Then you have to run hack/update-codegen.sh
and you get the new bits in there.

Reply:
Updated genreconciler with annotation filter. Now mismatching brokerclass won't get the finalizer.